spinnman/connections/scp_request_pipeline.py

```diff
@@ -224,69 +231,61 @@ def _single_retrieve(self, timeout):
 
                 # Remove the sequence from the outstanding responses
                 self._remove_record(seq)
-                self._in_progress -= 1
 
-    def _handle_receive_timeout(self, to_resend):
+    def _handle_receive_timeout(self):
         self._n_timeouts += 1
 
         # If there is a timeout, all packets remaining are resent
-        time_now = time.time()
         to_remove = list()
         for seq, request_sent in self._requests.iteritems():
-            if time_now - self._send_time[seq] >= self._packet_timeout:
-                to_resend.append((seq, request_sent))
-                self._in_progress -= 1
+            self._in_progress -= 1
+            try:
+                self._resend(seq, request_sent, "timeout")
+            except Exception as e:
+                self._error_callbacks[seq](
+                    request_sent, e, sys.exc_info()[2])
                 to_remove.append(seq)
+
         for seq in to_remove:
-            del self._requests[seq]
+            self._remove_record(seq)
 
-    def _resend(self, seq, request_sent):
+    def _resend(self, seq, request_sent, reason):
         if self._retries[seq] > 0:
```

Comment:
I notice that we're still using inverted test logic here. It's better to write “if bad-condition: raise exception;<newline>good-thing” instead of “if not bad-condition: good-thing else: raise exception”.